drivers/lora/lora_basics_modem/lbm_common.c

```diff
@@ -0,0 +1,549 @@
+				   pkt_status.snr_pkt_in_db, data->rx_state.async.user_data);
+}
+
+static void op_done_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct lbm_lora_data_common *data =
+		CONTAINER_OF(dwork, struct lbm_lora_data_common, op_done_work);
+	const struct device *dev = data->dev;
+	const struct lbm_lora_config_common *config = dev->config;
+	struct k_poll_signal *sig_done;
+	ral_irq_t irq_state;
+	ral_status_t status;
+	bool release = false;
+	bool error_irq;
+	int ret = 0;
+
+	LOG_DBG("%s: %d", __func__, data->modem_mode);
+
+	switch (data->modem_mode) {
+	case MODE_SLEEP:
+		LOG_WRN("Unexpected modem mode");
+		return;
+	case MODE_TX:
+	case MODE_CW:
+		status = ral_handle_tx_done(&config->ralf.ral);
+		if (status != RAL_STATUS_OK) {
+			LOG_WRN("RAL handle TX done failed (%d)", status);
+		}
+		release = true;
+		break;
+	case MODE_RX:
+		status = ral_handle_rx_done(&config->ralf.ral);
+		if (status != RAL_STATUS_OK) {
+			LOG_WRN("RAL handle RX done failed (%d)", status);
+		}
+		ret = op_done_sync_rx(dev);
+		release = true;
+		break;
+	case MODE_RX_ASYNC:
+		status = ral_handle_rx_done(&config->ralf.ral);
+		if (status != RAL_STATUS_OK) {
+			LOG_WRN("RAL handle RX done failed (%d)", status);
+		}
+		op_done_async_rx(dev);
+		/* Don't release the modem here, RX continues */
+		break;
+	case MODE_CAD:
+		LOG_DBG("CAD complete (TBC)");
+		break;
+	}
+
+	/* Get and reset the current IRQ state */
+	(void)ral_get_irq_status(&config->ralf.ral, &irq_state);
+	(void)ral_clear_irq_status(&config->ralf.ral, RAL_IRQ_ALL);
+	error_irq = irq_state & (RAL_IRQ_RX_TIMEOUT | RAL_IRQ_RX_HDR_ERROR | RAL_IRQ_RX_CRC_ERROR);
+
+	/* Release the modem before running the user callback so that the notified thread can
+	 * immediately start another operation before the work item terminates. This requires
+	 * preserving the operation_done pointer, since modem_release clears it.
+	 */
+	sig_done = data->operation_done;
+
+	/* Modem should return to idle */
+	if (release) {
+		/* Return to sleep mode */
+		modem_release(dev);
+	}
+
+	/* Notify user that operation has completed */
+	if (sig_done) {
+		k_poll_signal_raise(sig_done, error_irq ? -EAGAIN : ret);
+	}
+}
+
+int lbm_lora_common_init(const struct device *dev)
+{
+	const struct lbm_lora_config_common *config = dev->config;
+	struct lbm_lora_data_common *data = dev->data;
+	ral_status_t status;
+
+	data->dev = dev;
+	k_work_init_delayable(&data->op_done_work, op_done_work_handler);
+	atomic_clear(&data->modem_state);
+
+	/* Initialise the radio abstraction layer */
+	status = ral_init(&config->ralf.ral);
+	if (status != RAL_STATUS_OK) {
+		LOG_ERR("RAL init failure (%d)", status);
+		return -EIO;
+	}
+
+	/* Enable all relevant interrupts */
+	status = ral_set_dio_irq_params(&config->ralf.ral,
+					RAL_IRQ_TX_DONE | RAL_IRQ_RX_DONE | RAL_IRQ_RX_HDR_ERROR |
+						RAL_IRQ_RX_CRC_ERROR | RAL_IRQ_CAD_DONE |
+						RAL_IRQ_CAD_OK);
```

Comment:
indentation looks off, you may want to consider having these in a define as well maybe